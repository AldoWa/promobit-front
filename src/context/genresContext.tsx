import { createContext, useCallback, useEffect, useState } from 'react';

import api from '../services/api';

interface GenresProviderProps {
  children: React.ReactNode;
}

interface Genre {
  id: number;
  name: string;
}

export interface GenresResponse {
  genres: Genre[];
}

export interface GenresContextData {
  genres: Genre[];
  handleGenre: (id: number, type: 'add' | 'remove') => void;
  genresIDsSelected: number[];
  resetGenresSelected: () => void;
}

export const GenresContext = createContext<GenresContextData>(
  {} as GenresContextData
);
export function GenresProvider({ children }: GenresProviderProps) {
  const [genres, setGenres] = useState<Genre[]>([]);
  const [genresIDsSelected, setGenresSelected] = useState<number[]>([]);

  useEffect(() => {
    const getGenres = async () => {
      const { data } = await api.get<GenresResponse>('genre/movie/list?language=pt')

      setGenres(data.genres)
    }

    getGenres()
  }, [])

  const addGenreSelected = useCallback((id: number) => {
    setGenresSelected([...genresIDsSelected, id])
  }, [genresIDsSelected])

  const removeGenreSelected = useCallback((id: number) => {
    const newGenres = genresIDsSelected.filter(genre => genre !== id)
    setGenresSelected(newGenres)
  }, [genresIDsSelected])

  const resetGenresSelected = useCallback(() => setGenresSelected([]), [])

  const handleGenre = useCallback((id: number, type: 'add' | 'remove') => {
    if (type === 'add') {
      addGenreSelected(id)
      return;
    }
    removeGenreSelected(id)
  }, [addGenreSelected, removeGenreSelected])


  return (
    <GenresContext.Provider value={{ genres, genresIDsSelected, handleGenre, resetGenresSelected }}>
      {children}
    </GenresContext.Provider>
  );
}
