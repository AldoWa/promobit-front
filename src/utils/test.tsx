import { RenderOptions, render } from '@testing-library/react'

import { ChakraProvider } from '@chakra-ui/react'

interface AllTheProvidersProps {
  children: React.ReactNode
}

const AllTheProviders = ({children}: AllTheProvidersProps) => {
  return (
    <ChakraProvider>
      {children}
    </ChakraProvider>
  )
}

const customRender = (ui:React.ReactElement, options?: RenderOptions) =>
  render(ui, {wrapper: AllTheProviders, ...options})

// eslint-disable-next-line react-refresh/only-export-components
export * from '@testing-library/react'

// override render method
export {customRender as render}