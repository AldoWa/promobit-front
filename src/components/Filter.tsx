import React, { useContext } from 'react';
import Genres from './Genres';
import { HStack, Text, VStack } from '@chakra-ui/react';
import { GenresContext } from '../context/genresContext';

interface Genres {
  id: number;
  name: string;
}

const Filter: React.FC = () => {
  const { genres, handleGenre } = useContext(GenresContext);

  return (
    <VStack spacing={4}>
      <Text fontSize='0.875rem' fontWeight={700} color='white' data-testid="filter-text">FILTRE POR:</Text>
      <HStack flexWrap='wrap' justifyContent='center' data-testid="filter-genres">
        {genres.map(genre => (
          <Genres
            key={genre.id}
            text={genre.name}
            id={genre.id}
            handleGenre={handleGenre}
          />
        ))}
      </HStack>
    </VStack>
  );
}

export default Filter;
