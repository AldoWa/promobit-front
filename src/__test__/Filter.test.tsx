import { GenresContextData } from '../context/genresContext';
import { render, screen } from '../utils/test';
import Filter from '../components/Filter';

jest.mock('../services/api.ts', () => {
	return {
    AUTHORIZATION: 'XXXX'
  }
})

describe('Filter', () => {
    const providerProps: GenresContextData = {
      genres: [
        {
          id: 28,
          name: 'Action'
        },
        {
          id: 29,
          name: 'Drama'
        }
      ],
      handleGenre: () => {},
      genresIDsSelected: [],
      resetGenresSelected: () => {}
    }

    it("Should render the filter with 2 genres", () => {
      render(<Filter />, { providerProps })
      
      const filterByText = screen.getByTestId('filter-text')

      expect(filterByText).toBeInTheDocument()
      expect(filterByText).toHaveTextContent('FILTRE POR:');

      const element = screen.getAllByTestId('genre')
      expect(element.length).toBe(2)
    });
    
    it("Should match snapshot", () => {
      const { container } = render(<Filter />, { providerProps })
      expect(container).toMatchSnapshot();
    });
});