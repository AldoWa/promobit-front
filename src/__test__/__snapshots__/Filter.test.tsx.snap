// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Should match snapshot 1`] = `
<div>
  <div
    class="chakra-stack css-o51fhk"
  >
    <p
      class="chakra-text css-49y51g"
      data-testid="filter-text"
    >
      FILTRE POR:
    </p>
    <div
      class="chakra-stack css-1b1m5zm"
      data-testid="filter-genres"
    >
      <button
        class="css-1ru2wq5"
        data-testid="genre"
      >
        <p
          class="chakra-text css-1544uh1"
          data-testid="genre-text"
        >
          Action
        </p>
      </button>
      <button
        class="css-1ru2wq5"
        data-testid="genre"
      >
        <p
          class="chakra-text css-1544uh1"
          data-testid="genre-text"
        >
          Drama
        </p>
      </button>
    </div>
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
